\chapter[g2f: An R package for find and fill gaps in metabolic networks.]{Building a tissue-specific metabolic reconstruction: Finding and filling gaps in metabolic networks through `g2f' R package.}
\begin{tabular}{rm{12.5cm}}
\textsf{\textbf{Original title:}}& g2f: An R package for find and fill gaps in metabolic networks.\\
\textsf{\textbf{Written by:}} & \textit{Kelly Botero, Daniel Osorio, Janneth Gonzalez and Andrés Pinzón-Velasco}\\ 
\end{tabular}
\section*{Abstract}
The g2f package was designed as a tool to find the gaps (metabolites not produced or not consumed in any other reaction), and fill it from the stoichiometric reactions of a reference metabolic reconstruction using a weighting function. Also the option to download all the set of gene-associated stoichiometric reactions for a specific organism from the KEGG database is available.
\section{Introduction}
Genome-scale metabolic network reconstructions (GMNR) specify the chemical reactions catalyzed by hundreds of enzymes (registered in enzyme commission – E.C.) and cover the molecular function of a substantial fraction of a genome [1]. The main goal of these network reconstructions is to relate the genome of a given organism with its physiology, incorporating every metabolic transformation that this organism can perform [2], [3]. The GMNR are converted into computational models for simulation of metabolism and gain insight into the complex interactions that give rise to the metabolic capabilities [4], [5]. The predictive accuracy of a model depends on the comprehensiveness and biochemical fidelity of the reconstruction [6]. The GMNR construction process can be synthesized into two fundamental stages: (1) draft network reconstruction, here the reactions associated with the enzymes that participate in the metabolism of a particular organism, are downloaded from specialized genome, biochemical and metabolic databases; and (2) refinement of the network manually or using computational algorithms. Similar steps are performed during the construction of a tissue-specific metabolic reconstruction, defined as subsets of reactions included in a genome-scale metabolic reconstruction that are highly associated with  the metabolism of a specific tissue. They are constructed from measured gene expression or proteomic data and permit characterize or predict the metabolic behavior for each tissue under any physiological condition. Due to only the reactions associated with an enzyme or gene can be mapped from the measured data, the spontaneous reactions, and non-facilitated diffusion are missing in first stages of a tissue-specific reconstruction.\\

The refinement stage of the reconstruction is a process to restore the connectivity network, where network gaps in the draft reconstruction are identified, and candidate reactions to fill the gap are find in literature and databases [7], [8]. Since the network reconstructions typically involve thousands of metabolic reactions, the refinement of them can be a very complex task [3].The network gaps can be associated with dead-end metabolites which cannot be imported/produced by any of the reactions in the network; or metabolites that are not used as substrates or released by any of the reactions in the network. When the metabolic network is transformed into a metabolic steady-state model to optimize the distribution of metabolic flux under an objective function, the presence of this type of metabolites can be problematic, due to the flux cannot pass through them due to the incomplete connectivity with the rest of the network [8].  In a high-quality model, all reactions should be able to carry flux if all relevant exchange reactions are available [3]. The lack of flux in dead-end metabolites is propagated downstream/upstream, depending if the metabolites are not produced or not consumed, giving rise to additional metabolites that cannot carry any flux [8]. This can block out the net flux of the objective function and when it is not blocked, bias in the biological conclusions increase. The manual refinement is an iterative process to assemble a higher confidence compendium of organism-specific metabolism in a draft metabolic network reconstruction [9]–[11]. This type refinement requires time and a labor intensive of use of available literature, databases and experimental data [11], [12]. For example, if the genome annotation of the target organism is present in KEGG [13], reactions associated with the genes can be identified in KEGG maps and track dead-end metabolites under the organism-specific metabolic environment [7]. Given GMNR account for hundreds or thousands of biochemical reactions, this task is very complex and can lead to both, the introduction of new errors and to overlook some others.\\

Network gap refinement also can be performed using implemented algorithms to identify blocked reactions through the optimization of the metabolic model, and to fill metabolic functions gaps, by means of the search of reactions associated with the missing functions in available databases [3], [6]. However, this approach only fill gaps for a single biological objective function; other implemented approaches based on the identification and filling of dead-ends metabolites included in GAMS [8] and COBRA [14] packages, allows fill the networks gaps for all metabolism, generating a list of reactions that permit restore the network connectivity and eliminate blocked reactions that no carry flux in a model steady state [11]. However, implemented COBRA [14] gap fill algorithm operate as tools under the commercial  MATLAB® environment and GAMS [8] only allow the gap fill using MetaCyc databases as reference [15].\\

With the aim of offering an open source tool that facilitates the refinement and depuration of drafts network reconstructions and metabolic models, we introduce the ‘g2f’ R package. It includes five functions to identify and fill gaps, as well as, to depurate metabolic networks of blocked reactions (no activated under any scenario) and a function to associate GPR to reactions included in a metabolic reconstruction using the KEGG database as a reference. To identify network gaps in a metabolic network, two mechanisms were implemented: (1) Identification of the dead-end metabolites based on the stoichiometric matrix, and (2) the identification of all blocked through an iterative process of model optimization under all possible objective functions (each reaction by optimization). To gap fill, we identify by metabolite mapping a set of candidate reactions to restore the connectivity network from a reference metabolic database. Reference could be user-provided or created from the KEGG database [13], [16] for a customized organism through an implemented function.
\section{Installation and functions}
\texttt{g2f} includes four functions and is available for download and installation from CRAN, the
Comprehensive R Archive Network. To install and load it, just type:
\begin{Schunk}
\begin{Sinput}
> install.packages("g2f")
> library(g2f)
\end{Sinput}
\end{Schunk}
The \texttt{g2f} package requires R version 2.10 or higher. Development releases of the package are available on the GitHub repository \texttt{http://github.com/gibbslab/g2f}.
\subsection{Downloading a reference from the KEGG database}
\subsection{Calculating the addition cost}
\subsection{Performing a gap find and fill}
\subsection{Identifying blocked reactions}

\section{Summary}